source/extensions/filters/http/decompressor/decompressor_filter.cc

```diff
@@ -0,0 +1,195 @@
+    const envoy::extensions::filters::http::decompressor::v3::Decompressor::CommonDirectionConfig&
+        proto_config,
+    const std::string& stats_prefix, Stats::Scope& scope, Runtime::Loader& runtime)
+    : stats_(generateStats(stats_prefix, scope)),
+      decompression_enabled_(proto_config.enabled(), runtime) {}
+
+DecompressorFilterConfig::RequestDirectionConfig::RequestDirectionConfig(
+    const envoy::extensions::filters::http::decompressor::v3::Decompressor::RequestDirectionConfig&
+        proto_config,
+    const std::string& stats_prefix, Stats::Scope& scope, Runtime::Loader& runtime)
+    : DirectionConfig(proto_config.common_config(), stats_prefix + "request.", scope, runtime),
+      advertise_accept_encoding_(
+          PROTOBUF_GET_WRAPPED_OR_DEFAULT(proto_config, advertise_accept_encoding, true)) {}
+
+DecompressorFilterConfig::ResponseDirectionConfig::ResponseDirectionConfig(
+    const envoy::extensions::filters::http::decompressor::v3::Decompressor::ResponseDirectionConfig&
+        proto_config,
+    const std::string& stats_prefix, Stats::Scope& scope, Runtime::Loader& runtime)
+    : DirectionConfig(proto_config.common_config(), stats_prefix + "response.", scope, runtime) {}
+
+DecompressorFilter::DecompressorFilter(DecompressorFilterConfigSharedPtr config)
+    : config_(std::move(config)) {}
+
+Http::FilterHeadersStatus DecompressorFilter::decodeHeaders(Http::RequestHeaderMap& headers,
+                                                            bool end_stream) {
+  // Headers only request, continue.
+  if (end_stream) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+  ENVOY_STREAM_LOG(debug, "DecompressorFilter::decodeHeaders: {}", *decoder_callbacks_, headers);
+
+  // Two responsibilities on the request side:
+  //   1. If response decompression is enabled (and advertisement is enabled), then advertise to
+  //      the upstream that this hop is able to decompress responses via the Accept-Encoding header.
+  if (config_->responseDirectionConfig().decompressionEnabled()) {
+    if (config_->requestDirectionConfig().advertiseAcceptEncoding()) {
+      headers.appendAcceptEncoding(config_->contentEncoding(), ",");
+      ENVOY_STREAM_LOG(
+          debug, "DecompressorFilter::decodeHeaders advertise Accept-Encoding with value '{}'",
+          *decoder_callbacks_, headers.AcceptEncoding()->value().getStringView());
+    }
+  }
+
+  //   2. If request decompression is enabled, then decompress the request.
+  return maybeInitDecompress(config_->requestDirectionConfig(), request_decompressor_,
+                             *decoder_callbacks_, headers);
+};
+
+Http::FilterDataStatus DecompressorFilter::decodeData(Buffer::Instance& data, bool) {
+  return maybeDecompress(
+      config_->requestDirectionConfig(),
+      request_decompressor_
+          ? Envoy::Compression::Decompressor::DecompressorOptRef(std::ref(*request_decompressor_))
+          : absl::nullopt,
+      *decoder_callbacks_, data);
+}
```

Comment:
Would it be simpler to have here something like this instead?
```c++
Http::FilterDataStatus DecompressorFilter::encodeData(Buffer::Instance& data, bool) {
    if (response_decompressor_ != nullptr) then {
        doDecompress(config_->requestDirectionConfig(), request_decompressor_, *decoder_callbacks_, data);
    }

    return  Http::FilterDataStatus::Continue;
}
```
Proper ownership and memory cleanup would be still there if `doCompress()` accepted a const reference to the unique pointer like in `maybeInitDescompress()`.